src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,259 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
     using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
     using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
 
+    /// <summary>
+    /// Implementation of Johnson's algorithm for finding all simple cycles in a graph.
+    /// A simple cycle is one where no node repeats in the cycle until getting back
+    /// to the beginning of the cycle.
+    ///
+    /// The algorithm uses Tarjan's algorithm for finding all strongly-connected
+    /// components in a graph. A strongly-connected component, or SCC, is a subgraph
+    /// in which all nodes can reach all other nodes in the subgraph.
+    ///
+    /// The algorithm begins by using Tarjan's algorithm to find the SCCs of the graph.
+    /// Each SCC is processed for cycles, using the smallest-id node of the SCC as the
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The algorithm has time complexity of O((n + e)(c + 1)) with n nodes, e edges,
+    /// and c elementary cycles.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
```

Comment:
>SccStack [](start = 55, length = 8)

Why is this a stack? Are you really using the properties of a stack here, or it could be a simple array? This is a question, not a request for a change yet.

Reply:
We are always only pushing and popping off of this data structure, in a manor indicative of a stack. Using an array would be more messy here, and not communicate as clearly the use of the data structure.

Reply:
Johnson algorithm requires processing components in the order of the smallest-numbered vertex. Stack structure (as used in this implementation) does not guarantee that.

---
In reply to: [404521809](https://github.com/microsoft/qsharp-compiler/pull/396#discussion_r404521809) [](ancestors = 404521809)

Reply:
Since there cannot be any cycles between SCCs, there should not be any reason that we could not process all of one SCC before moving on to the next SCC. Nodes within an SCC are still processed by their smallest numbered node first.

Reply:
Oh but you were referring to the order of the SCCs, not the nodes. In the Johnson's algorithm the ordering is important because it keeps track of which node to process with a global index. When they finish with a node, it is removed from the graph and the index in incremented. This implementation does not use a global index to keep track of the what node to process next. That information is provided with each SCC from Tarjan's algorithm. It can return the minimum node in each SCC with no extra time cost because it keeps track of that while processing each node.

I am taking advantage of the fact that this implementation of Tarjan's is used specifically for Johnson's, so I can tailor the output of Tarjan's to help with Johnson's. There is a minimal size increase in carrying the integer minimum for each SCC, but this gives us the possible optimization of processing the SCCs in parallel.

Reply:
Agree. returning minimums from Tarjan is a good idea. That one number is not a huge overhead, and is worth it.

---
In reply to: [405306351](https://github.com/microsoft/qsharp-compiler/pull/396#discussion_r405306351) [](ancestors = 405306351)